active_storage_base64.gemspec

```diff
@@ -10,9 +10,13 @@ Gem::Specification.new do |s|
   s.authors = ['Nicolas Fabre, Ricky Ricoch']
   s.license = 'MIT'
 
-  #Dependencies
-  s.add_dependency 'rails', '>= 5.2.0'
+  # Dependencies
+  s.add_dependency 'rails', '~> 5.2.0'
```

Comment:
with doing this, you are limiting the gem to rails versions `5.2.x` . I don't think we want that, right?

Reply:
Yes. My concern here starts when I analyzed the Rails gem. When I did `bundle open activestorage` I saw the code there was different to the code I had by downloading the [Rails project](https://github.com/rails/rails) from github. There were some differences both in the code and the module name (the module that adds the `has_one_attached` helper is called `model` in the rails project and it is called `macro` when I did bundle open), so I don't know whether this is because the functionality might change in the future, therefore I am limiting the version to 5.2.x till a new Rails version comes out.

Reply:
For example, this is the module code from when I do `bundle open`
file: `lib/active_storage/attached/macros.rb`
```ruby
# frozen_string_literal: true

module ActiveStorage
  # Provides the class-level DSL for declaring that an Active Record model has attached blobs.
  module Attached::Macros
    # Specifies the relation between a single attachment and the model.
    #
    #   class User < ActiveRecord::Base
    #     has_one_attached :avatar
    #   end
    #
    # There is no column defined on the model side, Active Storage takes
    # care of the mapping between your records and the attachment.
    #
    # To avoid N+1 queries, you can include the attached blobs in your query like so:
    #
    #   User.with_attached_avatar
    #
    # Under the covers, this relationship is implemented as a +has_one+ association to a
    # ActiveStorage::Attachment record and a +has_one-through+ association to a
    # ActiveStorage::Blob record. These associations are available as +avatar_attachment+
    # and +avatar_blob+. But you shouldn't need to work with these associations directly in
    # most circumstances.
    #
    # The system has been designed to having you go through the ActiveStorage::Attached::One
    # proxy that provides the dynamic proxy to the associations and factory methods, like +attach+.
    #
    # If the +:dependent+ option isn't set, the attachment will be purged
    # (i.e. destroyed) whenever the record is destroyed.
    def has_one_attached(name, dependent: :purge_later)
      class_eval <<-CODE, __FILE__, __LINE__ + 1
        def #{name}
          @active_storage_attached_#{name} ||= ActiveStorage::Attached::One.new("#{name}", self, dependent: #{dependent == :purge_later ? ":purge_later" : "false"})
        end

        def #{name}=(attachable)
          #{name}.attach(attachable)
        end
      CODE

      has_one :"#{name}_attachment", -> { where(name: name) }, class_name: "ActiveStorage::Attachment", as: :record, inverse_of: :record, dependent: false
      has_one :"#{name}_blob", through: :"#{name}_attachment", class_name: "ActiveStorage::Blob", source: :blob

      scope :"with_attached_#{name}", -> { includes("#{name}_attachment": :blob) }

      if dependent == :purge_later
        after_destroy_commit { public_send(name).purge_later }
      else
        before_destroy { public_send(name).detach }
      end
    end

    # Specifies the relation between multiple attachments and the model.
    #
    #   class Gallery < ActiveRecord::Base
    #     has_many_attached :photos
    #   end
    #
    # There are no columns defined on the model side, Active Storage takes
    # care of the mapping between your records and the attachments.
    #
    # To avoid N+1 queries, you can include the attached blobs in your query like so:
    #
    #   Gallery.where(user: Current.user).with_attached_photos
    #
    # Under the covers, this relationship is implemented as a +has_many+ association to a
    # ActiveStorage::Attachment record and a +has_many-through+ association to a
    # ActiveStorage::Blob record. These associations are available as +photos_attachments+
    # and +photos_blobs+. But you shouldn't need to work with these associations directly in
    # most circumstances.
    #
    # The system has been designed to having you go through the ActiveStorage::Attached::Many
    # proxy that provides the dynamic proxy to the associations and factory methods, like +#attach+.
    #
    # If the +:dependent+ option isn't set, all the attachments will be purged
    # (i.e. destroyed) whenever the record is destroyed.
    def has_many_attached(name, dependent: :purge_later)
      class_eval <<-CODE, __FILE__, __LINE__ + 1
        def #{name}
          @active_storage_attached_#{name} ||= ActiveStorage::Attached::Many.new("#{name}", self, dependent: #{dependent == :purge_later ? ":purge_later" : "false"})
        end

        def #{name}=(attachables)
          #{name}.attach(attachables)
        end
      CODE

      has_many :"#{name}_attachments", -> { where(name: name) }, as: :record, class_name: "ActiveStorage::Attachment", inverse_of: :record, dependent: false do
        def purge
          each(&:purge)
          reset
        end

        def purge_later
          each(&:purge_later)
          reset
        end
      end
      has_many :"#{name}_blobs", through: :"#{name}_attachments", class_name: "ActiveStorage::Blob", source: :blob

      scope :"with_attached_#{name}", -> { includes("#{name}_attachments": :blob) }

      if dependent == :purge_later
        after_destroy_commit { public_send(name).purge_later }
      else
        before_destroy { public_send(name).detach }
      end
    end
  end
end
```



and this is the code downloaded from github
File: `lib/active_storage/attached/model.rb`
```ruby
# frozen_string_literal: true

module ActiveStorage
  # Provides the class-level DSL for declaring an Active Record model's attachments.
  module Attached::Model
    extend ActiveSupport::Concern

    class_methods do
      # Specifies the relation between a single attachment and the model.
      #
      #   class User < ActiveRecord::Base
      #     has_one_attached :avatar
      #   end
      #
      # There is no column defined on the model side, Active Storage takes
      # care of the mapping between your records and the attachment.
      #
      # To avoid N+1 queries, you can include the attached blobs in your query like so:
      #
      #   User.with_attached_avatar
      #
      # Under the covers, this relationship is implemented as a +has_one+ association to a
      # ActiveStorage::Attachment record and a +has_one-through+ association to a
      # ActiveStorage::Blob record. These associations are available as +avatar_attachment+
      # and +avatar_blob+. But you shouldn't need to work with these associations directly in
      # most circumstances.
      #
      # The system has been designed to having you go through the ActiveStorage::Attached::One
      # proxy that provides the dynamic proxy to the associations and factory methods, like +attach+.
      #
      # If the +:dependent+ option isn't set, the attachment will be purged
      # (i.e. destroyed) whenever the record is destroyed.
      def has_one_attached(name, dependent: :purge_later)
        generated_association_methods.class_eval <<-CODE, __FILE__, __LINE__ + 1
          p 'before def name'
          def #{name}
            @active_storage_attached_#{name} ||= ActiveStorage::Attached::One.new("#{name}", self)
          end

          def #{name}=(attachable)
            attachment_changes["#{name}"] =
              if attachable.nil?
                ActiveStorage::Attached::Changes::DeleteOne.new("#{name}", self)
              else
                ActiveStorage::Attached::Changes::CreateOne.new("#{name}", self, attachable)
              end
          end
        CODE

        has_one :"#{name}_attachment", -> { where(name: name) }, class_name: "ActiveStorage::Attachment", as: :record, inverse_of: :record, dependent: :destroy
        has_one :"#{name}_blob", through: :"#{name}_attachment", class_name: "ActiveStorage::Blob", source: :blob

        scope :"with_attached_#{name}", -> { includes("#{name}_attachment": :blob) }

        after_save { attachment_changes[name.to_s]&.save }

        after_commit(on: %i[ create update ]) { attachment_changes.delete(name.to_s).try(:upload) }

        ActiveRecord::Reflection.add_attachment_reflection(
          self,
          name,
          ActiveRecord::Reflection.create(:has_one_attached, name, nil, { dependent: dependent }, self)
        )
      end

      # Specifies the relation between multiple attachments and the model.
      #
      #   class Gallery < ActiveRecord::Base
      #     has_many_attached :photos
      #   end
      #
      # There are no columns defined on the model side, Active Storage takes
      # care of the mapping between your records and the attachments.
      #
      # To avoid N+1 queries, you can include the attached blobs in your query like so:
      #
      #   Gallery.where(user: Current.user).with_attached_photos
      #
      # Under the covers, this relationship is implemented as a +has_many+ association to a
      # ActiveStorage::Attachment record and a +has_many-through+ association to a
      # ActiveStorage::Blob record. These associations are available as +photos_attachments+
      # and +photos_blobs+. But you shouldn't need to work with these associations directly in
      # most circumstances.
      #
      # The system has been designed to having you go through the ActiveStorage::Attached::Many
      # proxy that provides the dynamic proxy to the associations and factory methods, like +#attach+.
      #
      # If the +:dependent+ option isn't set, all the attachments will be purged
      # (i.e. destroyed) whenever the record is destroyed.
      def has_many_attached(name, dependent: :purge_later)
        generated_association_methods.class_eval <<-CODE, __FILE__, __LINE__ + 1
          def #{name}
            @active_storage_attached_#{name} ||= ActiveStorage::Attached::Many.new("#{name}", self)
          end

          def #{name}=(attachables)
            attachment_changes["#{name}"] =
              if attachables.nil? || Array(attachables).none?
                ActiveStorage::Attached::Changes::DeleteMany.new("#{name}", self)
              else
                ActiveStorage::Attached::Changes::CreateMany.new("#{name}", self, attachables)
              end
          end
        CODE

        has_many :"#{name}_attachments", -> { where(name: name) }, as: :record, class_name: "ActiveStorage::Attachment", inverse_of: :record, dependent: :destroy do
          def purge
            each(&:purge)
            reset
          end

          def purge_later
            each(&:purge_later)
            reset
          end
        end
        has_many :"#{name}_blobs", through: :"#{name}_attachments", class_name: "ActiveStorage::Blob", source: :blob

        scope :"with_attached_#{name}", -> { includes("#{name}_attachments": :blob) }

        after_save { attachment_changes[name.to_s]&.save }

        after_commit(on: %i[ create update ]) { attachment_changes.delete(name.to_s).try(:upload) }

        ActiveRecord::Reflection.add_attachment_reflection(
          self,
          name,
          ActiveRecord::Reflection.create(:has_many_attached, name, nil, { dependent: dependent }, self)
        )
      end
    end

    def attachment_changes #:nodoc:
      @attachment_changes ||= {}
    end

    def reload(*) #:nodoc:
      super.tap { @attachment_changes = nil }
    end
  end
end
```

Reply:
Ok that makes sense.. I guess we could think on some solutions that are less coupled to the rails implementation so we don't need to change the solution all the time